spring-cloud-stream-binder-kinesis-core/src/main/java/org/springframework/cloud/stream/binder/kinesis/provisioning/KinesisStreamProvisioner.java

```diff
@@ -55,19 +61,40 @@ public KinesisStreamProvisioner(AmazonKinesis amazonKinesis,
 	@Override
 	public ProducerDestination provisionProducerDestination(String name,
 			ExtendedProducerProperties<KinesisProducerProperties> properties) throws ProvisioningException {
+		logger.info("Using Kinesis stream for outbound: " + name);
 
-		KinesisProducerDestination producer = new KinesisProducerDestination(name);
+		KinesisStream stream = createOrUpdate(name, properties.getPartitionCount());
 
-		return producer;
+		return new KinesisProducerDestination(stream.getName(), stream.getShards());
 	}
 
 	@Override
 	public ConsumerDestination provisionConsumerDestination(String name, String group,
 			ExtendedConsumerProperties<KinesisConsumerProperties> properties) throws ProvisioningException {
+		logger.info("Using Kinesis stream for inbound: " + name);
+
+		int shardCount = properties.getInstanceCount() * properties.getConcurrency();
+
+		KinesisStream stream = createOrUpdate(name, shardCount);
+
+		return new KinesisConsumerDestination(stream.getName(), stream.getShards());
+	}
+
+	private KinesisStream createOrUpdate(String name, Integer shards) {
+		try {
+			DescribeStreamResult streamResult = amazonKinesis.describeStream(name);
+			logger.info("Stream found, using existing stream");
 
-		KinesisConsumerDestination consumer = new KinesisConsumerDestination(name);
+			return new KinesisStream(name, streamResult.getStreamDescription().getShards().size());
```

Comment:
Right, `createOrUpdate` was intended to house shard creation/manipulation eventually, but I intentionally wanted to limit the scope of this work to stream auto-creation. This is also why I didn't fix the integration tests dealing with Kinesalite that didn't seem to deal with this end-goal. I agree with your idea of an `updateShardCount` property, but similarly left it out to be part of another effort.

If you would like such shard handling and finishing of integration tests to be part of this effort let me know and I'll look at those as well.